pkg/binding/crd.go

```diff
@@ -35,19 +37,50 @@ type APIResource struct {
 }
 
 // Handle CRDs should account for CRDs being added or deleted to start/stop new informers as needed
-func (c *Controller) handleCRD(_ runtime.Object) error {
-	toStartList, toStopList, err := c.checkAPIResourcesForUpdates()
-	if err != nil {
-		return err
+func (c *Controller) handleCRD(obj runtime.Object) error {
+	uObj := obj.(*unstructured.Unstructured)
+	var crdObj *apiextensionsv1.CustomResourceDefinition
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(uObj.UnstructuredContent(), &crdObj); err != nil {
+		return fmt.Errorf("failed to convert Unstructured to CRD: %w", err)
+	}
+
+	// for each CustomResourceDefinitionVersion, follow a decision tree to tell whether the corresponding gvr should be watched
+	toStartList, toStopList := []APIResource{}, []string{}
+	for _, ver := range crdObj.Spec.Versions {
+		gvr := APIResource{
+			groupVersion: schema.GroupVersion{
+				Group:   crdObj.Spec.Group,
+				Version: ver.Name,
+			},
+			resource: metav1.APIResource{
+				Name: crdObj.Spec.Names.Plural,
+				Kind: crdObj.Spec.Names.Kind,
+			},
+		}
+		key := util.KeyForGroupVersionKind(gvr.groupVersion.Group, gvr.groupVersion.Version, gvr.resource.Kind)
+		if isBeingDeleted(obj) {
+			toStopList = append(toStopList, key)
+			continue
+		}
+		if !c.includedToWatch(gvr) {
+			toStopList = append(toStopList, key)
+			continue
+		}
+		if !ver.Served {
+			toStopList = append(toStopList, key)
+			continue
+		}
+		toStartList = append(toStartList, gvr)
```

Comment:
Welcome to the world of CRD dynamics. Please review https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/v1/types.go#L308 before proceeding further.